torch/_functorch/aot_autograd.py

```diff
@@ -138,86 +138,365 @@ def posthook(grad_input, grad_output):
+    #   (same as)
+    #   len(x for x in input_info if x.MutationType != MutationType.none) + len(output_info) + num_intermediate_bases
     # For every output *and* mutated input returned from the forward,
     # tells us whether or not the output should require gradients or not
-    requires_grad_out_info: List[bool]
-    # length: # fw outputs
-    aliased_output_info: List[OutputAliasInfo]
-
-def gen_alias_from_base(aliased_base_tensor, size, stride, storage_offset, target_meta_tensor):
-    # handle R2C and C2R
+    requires_grad_info: List[bool]
+    # length = len(x for x in input_info if x.MutationType != MutationType.none)
+    # This is pre-computed for perf: it contains the index of every element
+    # of input_info that corresponds to a mutation (data or metadata or both)
+    mutated_inp_indices: List[int]
+    # length = len(x for x in output_info if x.output_type != OutputType.non_alias)
+    # This is pre-computed for perf: it contains the index of every element
+    # of output_info that corresponds to an alias (either of an input or intermediate)
+    aliased_out_indices: List[int]
+    # length = the number of intermediate bases appended as outputs to the end of the forward graph.
+    # Note: this is not necessarily the same thing as:
+    #   len([x for x in output_info if x.output_type == OutputType.alias_of_intermediate])
+    # Because outputs might share a ._base, or an output's ._base might itself be
+    # another user output (in both cases, we won't redundantly append bases to the end of the graph)
+    num_intermediate_bases: int
+
+# This class exists because:
+# - the autograd.Function.forward() in aot autograd returns outputs that might alias inputs
+# - we only care about the metadata on those aliases, so we can regenerate them.
+#   We do not want them to participate in the autograd.Function.
+# We do that by wrapping them in an opaque class, so the autograd.Function
+# does not know to treat them as tensors.
+@dataclass(frozen=True)
+class TensorAlias:
+    alias: torch.Tensor
+    requires_grad: bool
+
+def has_same_metadata(t1, t2):
+    return t1.size() == t2.size() and t1.stride() == t2.stride() and t1.storage_offset() == t2.storage_offset()
+
+def is_data_mutation(m: MutationType):
+    # We should really turn on mypy
+    assert isinstance(m, MutationType)
+    return m in {MutationType.data, MutationType.data_and_metadata}
+
+def gen_alias_from_base(aliased_base_tensor, target_meta_tensor, target_requires_grad):
+    # Try to do view-replay if possible.
+    # fall back to .as_strided() if we can't.
+    if target_meta_tensor._base is not None:
+        # The base that we want to replay our view off of might have a different shape than the view's original base.
+        b = target_meta_tensor._base
+        reshaped_base_tensor = aliased_base_tensor.as_strided(b.size(), b.stride(), b.storage_offset())
+        if torch.Tensor._has_same_metadata(target_meta_tensor._base, reshaped_base_tensor):
```

Comment:
@albanD I made calling into the view-replay logic here best-effort (try to reshape our synthetic base to look like the alias's `._base`, then check if their metadata is the same). LMK what you think - I could also just update `._view_func()` to accept a flag so it returns None instead of erroring, so I don't need to bind a separate function to python.

Reply:
It seems way better to update `_view_func`. I'd also be fine with another function that is like `_has_view_func`. I have no idea what the heck the `_has_same_metadata` call above is doing, and it's not like you conserved binding a function to Python, you had to bind one anyway! Do you even need `_has_same_metadata` for anything else?